### PR TITLE
Client (server) tests

### PR DESCRIPTION
**Pull request by mprudnik**
- [x] tests and linter show no problems (`npm t`)
- [x] tests are added/updated for bug fixes and new features
- [x] code is properly formatted (`npm run fmt`)
- [ ] description of changes is added in CHANGELOG.md
- [ ] update .d.ts typings

### CODE REVIEW

**Comment by tshemsedinov** [comment]
```
       connection.on('message', (data, isBinary) => {
-        if (isBinary) this.binary(client, data);
+        if (isBinary) this.binary(client, new Uint8Array(data));
         else this.message(client, data);
       });
```

**Comment by mprudnik** [comment]
@tshemsedinov Tests are passing now.
Could you please review how the tests are written before landing this PR?
The idea was to use only client code to test it separately, create mock server for each case and run a basic success case scenario.

File: test/client.js [hunk 1]
```diff
@@ -1,3 +1,280 @@
...
+    await client.opening;
+    await client.load('test');
+  });
+
+  test.afterEach(async () => void client.close());
+
+  test.testAsync('handles simple api calls', async (t) => {
+    const result = await client.api.test.test();
+    t.strictEqual(result, { success: true });
+  });
```
Comment (tshemsedinov) [inline test/client.js:77]: I'd prefer `subtest` instead of `t`

**Review by tshemsedinov — APPROVED** [review]
Are we ready for landing? LGTM except a few naming issues

**Comment by mprudnik** [comment]
@tshemsedinov Yes, ready

File: test/client.js (hunk 1 above)
Comment (tshemsedinov) [inline test/client.js:77]:
```suggestion
  test.testAsync('handles simple api calls', async (subtest) => {
    const result = await client.api.test.test();
    subtest.strictEqual(result, { success: true });
  });
```

File: test/client.js [hunk 2]
```diff
@@ -1,3 +1,280 @@
...
+
+  test.testAsync('handles parallel api calls', async (t) => {
+    const promises = [];
+    for (let i = 0; i < 10; i++) promises.push(client.api.test.test());
+    const res = await Promise.all(promises);
+    for (const r of res) t.strictEqual(r, { success: true });
+  });
```
Comment (tshemsedinov) [inline test/client.js:84]:
```suggestion
  test.testAsync('handles parallel api calls', async (subtest) => {
    const promises = [];
    for (let i = 0; i < 10; i++) promises.push(client.api.test.test());
    const res = await Promise.all(promises);
    for (const r of res) subtest.strictEqual(r, { success: true });
  });
```

File: test/client.js [hunk 3]
```diff
@@ -1,3 +1,280 @@
...
+
+  test.testAsync('handles call timeouts', async (t) => {
+    await t.rejects(client.api.test.timeout(), new Error('Request timeout'));
+  });
+
+  test.testAsync('handles api errors', async (t) => {
+    await t.rejects(client.api.test.error(), new Error('Error message'));
+  });
+});
```
Comment (tshemsedinov) [inline test/client.js:93]:
```suggestion
  test.testAsync('handles call timeouts', async (subtest) => {
    await subtest.rejects(client.api.test.timeout(), new Error('Request timeout'));
  });

  test.testAsync('handles api errors', async (subtest) => {
    await subtest.rejects(client.api.test.error(), new Error('Error message'));
  });
});
```

File: test/client.js [hunk 4]
```diff
@@ -1,3 +1,280 @@
...
+    t.strictEqual(echoResult, { success: true });
+  });
+
+  test.testAsync('handles events from server', async (t) => {
+    const ping = await new Promise((resolve) =>
+      client.api.test.on('ping', resolve),
+    );
+    t.strictEqual(ping, { ping: true });
+  });
+});
```
Comment (tshemsedinov) [inline test/client.js:160]:
```suggestion
  test.testAsync('handles event emitting', async (subtest) => {
    client.api.test.emit('echo', { test: true });
    client.api.test.on('*', console.log);
    const echoResult = await new Promise((resolve) =>
      client.api.test.once('echo', resolve),
    );
    subtest.strictEqual(echoResult, { success: true });
  });

  test.testAsync('handles events from server', async (subtest) => {
    const ping = await new Promise((resolve) =>
      client.api.test.on('ping', resolve),
    );
    subtest.strictEqual(ping, { ping: true });
  });
});
```

File: test/client.js [hunk 5]
```diff
@@ -1,3 +1,280 @@
...
+  });
+
+  test.testAsync('handles file downloads', async (t) => {
+    const name = 'download-stream';
+    const { id } = await client.api.test.download({ name });
+    const readable = client.getStream(id);
+    const blob = await readable.toBlob();
+    const data = await blob.text();
+    t.strictEqual(data, 'Some random data for upload to the client');
+  });
```
Comment (tshemsedinov) [inline test/client.js:279]:
```suggestion
  test.testAsync('handles file uploades', async (subtest) => {
    const data = 'Some random data for upload to the server';
    const name = 'upload-stream';
    const blob = new Blob([data]);
    blob.name = name;
    const stream = client.createBlobUploader(blob);
    await stream.upload();
    const uploadedFile = await client.api.test.getStreamData({ id: stream.id });
    subtest.strictEqual(uploadedFile.name, name);
    subtest.strictEqual(uploadedFile.size, blob.size);
    subtest.strictEqual(uploadedFile.data.join(''), data);
    subtest.strictEqual(uploadedFile.status, 'ended');
  });

  test.testAsync('handles file downloads', async (subtest) => {
    const name = 'download-stream';
    const { id } = await client.api.test.download({ name });
    const readable = client.getStream(id);
    const blob = await readable.toBlob();
    const data = await blob.text();
    subtest.strictEqual(data, 'Some random data for upload to the client');
  });
```

File: test/client.js [hunk 6]
```diff
@@ -1,3 +1,280 @@
...
+
+  test.testAsync('handles parallel api calls', async (subtest) => {
+    const promises = [];
+    for (let i = 0; i < 10; i++) promises.push(client.api.test.test());
+    const res = await Promise.all(promises);
+    for (const r of res) subtest.strictEqual(r, { success: true });
+  });
+
+  test.testAsync('handles call timeouts', async (subtest) => {
+    await subtest.rejects(client.api.test.timeout(), new Error('Request timeout'));
```
Comment (tshemsedinov) [inline test/client.js:87]:
```suggestion
    const promise = client.api.test.timeout();
    await subtest.rejects(promise, new Error('Request timeout'));
```

**Comment by tshemsedinov** [comment]
@mprudnik Landed, Thanks!

**Pull request merged**